### ISSUE TITLE
All custom API calls return 404 in the last couple updates

### ISSUE THREAD

**Issue by eXistenZe** [2sic/2sxc#1830]
**I'm submitting a ...** 
[x] bug report 

**...about** 
[x] APIs like REST

**Current behavior**
I get this error in every single call I try to make:

template:
`$2sxc(@Dnn.Module.ModuleID).webApi.post("demo/returnDxInfo", {}, woundtype, true).then(handleState);`

API (api/demoController.cs):
```
[HttpPost]
    [DnnModuleAuthorize(AccessLevel = SecurityAccessLevel.Anonymous)]
    [ValidateAntiForgeryToken]
    public DxData returnDxInfo([FromBody]int dx)
    {
```

Response:
`{"Message":"No HTTP resource was found that matches the request URI 'https://site.pt/pt-pt/API/2sxc/app/auto/api/demo/returnDxInfo'.","MessageDetail":"No controller was created to handle this request."}`

**Expected behavior**
API works fine.
This is actually old code that worked fine until now.

* **2sxc version(s):** 10.06  
* **DNN:** [9.4]

**Comment by iJungleboy** [comment]
I'll check - didn't pop up yet

**Comment by eXistenZe** [comment]
Thanks. Nearly all my apps use custom controllers, so my site is on stanby until the fix :)
Nothing urgent nevertheless.

**Comment by iJungleboy** [comment]
Just fyi: on my setup 
* works: DNN 7.4.2 and 2sxc 10.06
* fails: dnn 9.4 RC1 and 2sxc 10.06

**Comment by iJungleboy** [comment]
On DNN 9.4 RC1 I get this error: 
```
<Error>
<Message>
2sxc Api Controller Finder: Error while selecting / compiling a controller for the request. Pls check the event-log and the code. See the inner exception for more details.
</Message>
<ExceptionMessage>
Object reference not set to an instance of an object.
</ExceptionMessage>
<ExceptionType>System.NullReferenceException</ExceptionType>
<StackTrace>
at ToSic.SexyContent.Environment.Dnn7.DnnInstanceInfo.get_Id() in C:\Projects\2sxc-dnn742\Website\DesktopModules\ToSIC_SexyContent\2sxc Dnn\Environment\Dnn7\DnnInstanceInfo.cs:line 12 at ToSic.SexyContent.ContentBlocks.ModuleContentBlock..ctor(IInstanceInfo instanceInfo, Log parentLog, ITenant tenant, IEnumerable`1 overrideParams) in C:\Projects\2sxc-dnn742\Website\DesktopModules\ToSIC_SexyContent\ToSic.Sxc\SexyContent\ContentBlocks\ModuleContentBlock.cs:line 36 at ToSic.SexyContent.WebApi.Helpers.GetSxcOfApiRequest(HttpRequestMessage request, Boolean allowNoContextFound, Log log) in C:\Projects\2sxc-dnn742\Website\DesktopModules\ToSIC_SexyContent\Sxc WebApi\Helpers.cs:line 52 at ToSic.SexyContent.WebApi.AppApiControllerSelector.SelectController(HttpRequestMessage request) in C:\Projects\2sxc-dnn742\Website\DesktopModules\ToSIC_SexyContent\Sxc WebApi\AppApiControllerSelector.cs:line 81
</StackTrace>
</Error>
```
Seem like something in DNN changed.

**Comment by iJungleboy** [comment]
I've spent some time on this, and it looks like 2sic is doing everything right, but DNN 9.4 does something differently. I'll try to ask for help of people who worked on DNN 9.4

**Comment by iJungleboy** [comment]
It's really a DNN issue - I found it and posted a fix, hope it gets added soon! See https://github.com/dnnsoftware/Dnn.Platform/issues/3045

**Comment by eXistenZe** [comment]
Thanks.
That was fast.

Is it some simple file or code replacing I need to do without compiling anything?

**Comment by iJungleboy** [comment]
Nope, DNN 9.4 really broke this. ATM we can only wait. I sent a pull request with the code that fixes it.

**Comment by iJungleboy** [comment]
Everybody - please vote on this on https://github.com/dnnsoftware/Dnn.Platform/issues/3045 because the team which decides on these changes never seem to have used this feature and is not aware of it's importance.

**Comment by eXistenZe** [comment]
How?

**Comment by iJungleboy** [comment]
Just go to https://github.com/dnnsoftware/Dnn.Platform/issues/3045 and add a thumbs-up, heart or many things :). just so people see that others actually need this.

**Comment by eXistenZe** [comment]
No luck. Do you think this will be fixed in the next weeks or should I revert to 9.3.2?

**Comment by iJungleboy** [comment]
The DNN team seems to fear breaking new things if they fix this. So they don't want to do this. Because of this, we'll fix it in 2sxc in 10.08. We already have a solution.

**Comment by iJungleboy** [comment]
I'll pre-release 10.8. Pleas test it :)

**Comment by nicpitsch** [comment]
Everything fine now with DNN 9.4.1 and 2sxc 10.08. Thanks for the effort.

Is your workaround based on a response you got on [your question concerning best practices  how to implement dynamically compiled webapis into DNN](https://github.com/dnnsoftware/Dnn.Platform/issues/3045#issuecomment-540957902)? 

I'm somehow concerned about the discussion in [DNN#3045](https://github.com/dnnsoftware/Dnn.Platform/issues/3045). I don't understand it tecnically, but it sounds like dynamic compiled code is considered a 'bad' thing?

**Comment by iJungleboy** [comment]
Hi @nicpitsch Dynamically compiled code is normal in asp.net - every MVC razor page, every webcontrol and many other things are dynamically compiled code. I believe the DNN guardians had never consider that WebApis could be done the same way, so they were very surprised to see this and did express concern because there need to be checks that only super-users can make changes to such files (just like with cshtml or ascx files). These checks are there, and we'll actively re-verify them to be sure. 

In terms of workaround that's fairly easy to do, and basically there never is a "can't be done" scenario, because this is what asp.net is all about. It would simply have been easier and IMHO better if it were solved at the core level, because otherwise 2sxc is the only plugin that can easily support dynamic web controllers.

**Comment by eXistenZe** [comment]
Upgraded to 9.4.1 and 10.08 and everything seems fine.

**Comment by iJungleboy** [comment]
awesome 👍

**Comment by mitchelsellers** [comment]
> > The DNN team seems to fear breaking new things if they fix this. So they don't want to do this. Because of this, we'll fix it in 2sxc in 10.08. We already have a solution.
> 
> How does that make sense? They changed things that broke your stuff. They need to fix it. I realize its more complicated than I can imagine, but still. #sigh (no response necessary)

Just to be clear on the decision point from the DNN Platform team, the changes that we introduced in 9.4.0 introduced improved functionality for the platform as a whole, and was done in a manner to limit any impacts to third parties.  

We have chosen to not support the requested change for 2sxc, as the changes in 9.4.0+ actually provide a more secure infrastructure for the entire DNN Platform community, which is the number one priority of the DNN Platform team.
